File: app/src/main/java/com/stm/evenote/rsync/opfactory/AbstractOperationFactory.java
package com.stm.evenote.rsync.opfactory;

import java.io.IOException;
import java.util.Collections;
import java.util.List;

import com.stm.evenote.rsync.model.Operation;
import com.stm.evenote.rsync.model.OperationFactory;
import com.stm.evenote.rsync.model.SyncFile;
import com.stm.evenote.rsync.model.SyncPath;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

abstract class AbstractOperationFactory implements OperationFactory {

  private static final Logger logger = LoggerFactory.getLogger(AbstractOperationFactory.class);

  protected final String workingDirectory;
  protected final Boolean delete;
  protected final List<String> excludeDeletionRegExps;

  AbstractOperationFactory(String workingDirectory, Boolean delete, List<String> excludeDeleteRegExps) {
    this.workingDirectory = workingDirectory;
    this.delete = delete;
    this.excludeDeletionRegExps = excludeDeleteRegExps == null ? Collections.emptyList() : excludeDeleteRegExps;
  }

  public Operation noOp(SyncFile file) {

    return () -> logger.debug("No operation for file '{}'", absolutePath(file));
  }

  public Operation updateFile(SyncFile fileToUpdate, SyncFile fileWithNewContent) {

    return () -> {
      var absoluteSyncPath = absolutePath(fileToUpdate);
      try {
        execUpdateFile(fileWithNewContent, absoluteSyncPath);
      } catch (IOException e) {
        logger.error("Could not update file '{}' with new data", absoluteSyncPath, e);
      }
    };
  }

  public Operation delete(SyncFile syncPathToDelete) {

    return () -> {
      var matchedExcludePattern = this.excludeDeletionRegExps.stream()
        .filter(syncPathToDelete::matches)
        .findFirst();

      var absoluteFile = absolutePath(syncPathToDelete);
      if (this.delete) {
        if (absoluteFile.isFile()) {
          if (matchedExcludePattern.isEmpty()) {
            execDeleteFile(absoluteFile);
          } else {
            logger.info("File '{}' matched exclude pattern '{}'. Won't delete.",
              absoluteFile, matchedExcludePattern.get());
          }
        } else if (absoluteFile.isEmptyDir()) {
          if (matchedExcludePattern.isEmpty()) {
            execDeleteDir(absoluteFile);
          } else {
            logger.info("Empty directory '{}' matches '{}'. Won't delete.",
              absoluteFile, matchedExcludePattern.get());
          }
        }
      } else {
        logger.info(
          "Deletion inactive. Use 'delete' option to delete file '{}'.",
          absoluteFile
        );
      }
    };
  }

  public Operation newFile(SyncFile newFile) {

    return () -> {
      var absoluteSyncPath = absolutePath(newFile);
      try {
        execNewFile(newFile, absoluteSyncPath);
      } catch (Exception e) {
        logger.error("Could not create new file '{}'", absoluteSyncPath, e);
      }
    };
  }

  protected abstract void execDeleteDir(SyncPath absoluteFile);

  protected abstract void execDeleteFile(SyncPath absoluteFile);

  protected abstract void execNewFile(SyncFile newFile, SyncPath absoluteSyncPath) throws Exception;

  protected abstract void execUpdateFile(SyncFile fileWithNewContent, SyncPath absoluteSyncPath) throws IOException;

  private SyncPath absolutePath(SyncFile fileToUpdate) {
    return fileToUpdate.getFilename().prepend(this.workingDirectory);
  }
}

File: app/src/main/kotlin/com/stm/evenote/rsync/opfactory/AbstractOperationFactory.kt
package com.stm.evenote.rsync.opfactory

import com.stm.evenote.rsync.model.Operation
import com.stm.evenote.rsync.model.OperationFactory
import com.stm.evenote.rsync.model.SyncFile
import com.stm.evenote.rsync.model.SyncPath
import org.slf4j.LoggerFactory
import java.io.IOException

abstract class AbstractOperationFactory(
    protected val workingDirectory: String,
    protected val delete: Boolean,
    protected val excludeDeletionRegExps: List<String>
) : OperationFactory {

    private val logger = LoggerFactory.getLogger(AbstractOperationFactory::class.java)

    override fun noOp(file: SyncFile): Operation {
        return Operation { logger.debug("No operation for file '${absolutePath(file)}'") }
    }

    override fun updateFile(fileToUpdate: SyncFile, fileWithNewContent: SyncFile): Operation {
        return Operation {
            val absoluteSyncPath = absolutePath(fileToUpdate)
            try {
                execUpdateFile(fileWithNewContent, absoluteSyncPath)
            } catch (e: IOException) {
                logger.error("Could not update file '${absoluteSyncPath}' with new data", e)
            }
        }
    }

    override fun delete(fileToDelete: SyncFile): Operation {
        return Operation {
            val matchedExcludePattern = excludeDeletionRegExps.stream()
                .filter(fileToDelete::matches)
                .findFirst()
            val absoluteFile = absolutePath(fileToDelete)
            if (delete) {
                if (absoluteFile.isFile) {
                    if (matchedExcludePattern.isEmpty) {
                        execDeleteFile(absoluteFile)
                    } else {
                        logger.info(
                            "File '${absoluteFile}' matched exclude pattern '${matchedExcludePattern.get()}'. Won't delete."
                        )
                    }
                } else if (absoluteFile.isEmptyDir) {
                    if (matchedExcludePattern.isEmpty) {
                        execDeleteDir(absoluteFile)
                    } else {
                        logger.info(
                            "Empty directory '${absoluteFile}' matches '${matchedExcludePattern.get()}'. Won't delete.",
                        )
                    }
                }
            } else {
                logger.info(
                    "Deletion inactive. Use 'delete' option to delete file '${absoluteFile}'.",

                    )
            }
        }
    }

    override fun newFile(newFile: SyncFile): Operation {
        return Operation {
            val absoluteSyncPath = absolutePath(newFile)
            try {
                execNewFile(newFile, absoluteSyncPath)
            } catch (e: Exception) {
                logger.error("Could not create new file '${absoluteSyncPath}'", e)
            }
        }
    }

    protected abstract fun execDeleteDir(absoluteFile: SyncPath)

    protected abstract fun execDeleteFile(absoluteFile: SyncPath)

    protected abstract fun execNewFile(newFile: SyncFile, absoluteSyncPath: SyncPath)

    protected abstract fun execUpdateFile(fileWithNewContent: SyncFile, absoluteSyncPath: SyncPath)

    private fun absolutePath(fileToUpdate: SyncFile): SyncPath {
        return fileToUpdate.filename.prepend(workingDirectory)
    }
}